Unable to build universal deploy for MacOS 🐛

### Initial notes from Sean: 

When `Arm64` mac runners became the default for GitHub it seems it broke our desktop deploys to MacOS by having only `Arm64` arch compatible executables get output as part of the release process. This can be fixed by changing the build options in the `package.json` to target `x86` as a separate deploy or by building a `universal` version. However, this is where the trouble really begins. It seems like for some reason the `universal` and `x86` builds that build on GitHub don't actually produce an `x86` compatible executable, instead I get this error:

<img width="984" alt="image" src="https://github.com/user-attachments/assets/5a0d18bc-2074-4c34-a0d5-4c18ce712285" />

This seems to be an error many others have gotten as well and the routes and various fixes they perform are varied. What is more annoying is it seems the File Upload App does exactly what we do [on my working branch](https://github.com/AllenInstitute/aics-fms-file-explorer-app/compare/main...bugfix/make-macos-build-universal?expand=1) but without the error ([see here](https://github.com/aics-int/aics-file-upload-app/blob/main/package.json#L20-L25). 

Related Issues
* https://github.com/electron/universal/issues/41
* https://github.com/electron-userland/electron-builder/issues/7512
* https://github.com/electron-userland/electron-builder/pull/6983
* https://github.com/electron-userland/electron-builder/issues/6433

Things I tried to so far:
* Upgrading electron & electron-builder
* Upgrading electron-builder to a middle ground version like 22
* Setting `mergeAsar` to false
* Setting the node modules included to exclude certain extensions
(various combos of above, not to say any of these should necessarily be ruled out since it may require the right combo)

### Update: Notes from Anya
I eventually narrowed down the issue to DuckDB. When I remove duckdb as a dependency completely (and noop all duckdb functionality) the universal builds work for both arch. I *think* they've fixed the mac incompatibility issue in 1.0.0 (we're using 0.8.x) BUT when I try to upgrade to 1.0.0, I get a whole bunch of nasty mapbox/node-pre-gyp dependency & webpack errors. Unfortunately, node-pre-gyp is no longer maintained, so I haven't had much luck in finding solutions
Related discussion: 
* https://github.com/duckdb/duckdb/discussions/7067
* https://github.com/mapbox/node-pre-gyp/issues/661
* This is for an older version but vaguely related? https://github.com/duckdb/duckdb/issues/5605
* https://github.com/duckdb/duckdb-node/issues/115

Currently the deployed versions of the desktop app are based on main, but built on their respective chips and manually added to the release.

Revisited this with duckdb 1.1.3, still seeing the same mapbox/node-pre-gyp errors:
`Module not found: Error: Can't resolve 'npm' in '<local/path/to/bff>/node_modules/@mapbox/node-pre-gyp/lib/util'`
Still unable to find a clear solution, so will still need separate executables for mac built on their respective chips

--- a/.github/workflows/manual-build.yml
+++ b/.github/workflows/manual-build.yml
@@ -14,6 +14,7 @@ on:
       mac_arch:
         description: "Target Mac Architecture (only used for macos-latest)"
         required: false
+        default: arm64
         type: choice
         options:
           - x64
@@ -32,6 +33,36 @@ jobs:
         with:
           node-version: 18
 
+      # For macOS builds, use one of two steps based on the chosen architecture.
+      - name: Build/release Electron app (mac x64)
+        if: ${{ github.event.inputs.os == 'macos-latest' && github.event.inputs.mac_arch == 'x64' }}
+        uses: AllenCellSoftware/action-electron-builder@fms-file-explorer
+        env:
+          AMPLITUDE_API_KEY: ${{ secrets.AMPLITUDE_API_KEY }}
+        with:
+          github_token: ${{ secrets.github_token }}
+          package_root: "packages/desktop"
+          release: false
+          # Override configuration to build only a dmg file for Intel (x64)
+          config_overrides: '{"mac": {"target": [{"target": "dmg", "arch": ["x64"]}]}}'
+          windows_certs: ${{ secrets.CSC_LINK }}
+          windows_certs_password: ${{ secrets.CSC_KEY_PASSWORD }}
+
+      - name: Build/release Electron app (mac arm64)
+        if: ${{ github.event.inputs.os == 'macos-latest' && github.event.inputs.mac_arch == 'arm64' }}
+        uses: AllenCellSoftware/action-electron-builder@fms-file-explorer
+        env:
+          AMPLITUDE_API_KEY: ${{ secrets.AMPLITUDE_API_KEY }}
+        with:
+          github_token: ${{ secrets.github_token }}
+          package_root: "packages/desktop"
+          release: false
+          # Override configuration to build only a dmg file for Apple Silicon (arm64)
+          config_overrides: '{"mac": {"target": [{"target": "dmg", "arch": ["arm64"]}]}}'
+          windows_certs: ${{ secrets.CSC_LINK }}
+          windows_certs_password: ${{ secrets.CSC_KEY_PASSWORD }}
+
+      # For non-macOS builds, use the normal build step.
       - name: Build/release Electron app (non-mac)
         if: ${{ github.event.inputs.os != 'macos-latest' }}
         uses: AllenCellSoftware/action-electron-builder@fms-file-explorer
@@ -44,33 +75,19 @@ jobs:
           windows_certs: ${{ secrets.CSC_LINK }}
           windows_certs_password: ${{ secrets.CSC_KEY_PASSWORD }}
 
-      - name: Build/release Electron app (mac)
+      # Upload artifacts appropriately.
+      - name: Upload build artifacts (mac)
         if: ${{ github.event.inputs.os == 'macos-latest' }}
-        uses: AllenCellSoftware/action-electron-builder@fms-file-explorer
-        env:
-          AMPLITUDE_API_KEY: ${{ secrets.AMPLITUDE_API_KEY }}
+        uses: actions/upload-artifact@v4
         with:
-          github_token: ${{ secrets.github_token }}
-          package_root: "packages/desktop"
-          release: false
-          # Override configuration to build only mac with the chosen architecture.
-          config_overrides: >
-            {"mac": {"target": "default", "arch": "${{ github.event.inputs.mac_arch }}"}, "linux": null, "win": null}
-          windows_certs: ${{ secrets.CSC_LINK }}
-          windows_certs_password: ${{ secrets.CSC_KEY_PASSWORD }}
+          name: electron-build-macos-${{ github.event.inputs.mac_arch }}
+          path: packages/desktop/build
+          retention-days: 7
 
       - name: Upload build artifacts (non-mac)
         if: ${{ github.event.inputs.os != 'macos-latest' }}
         uses: actions/upload-artifact@v4
         with:
           name: electron-build-${{ github.event.inputs.os }}
-          path: packages/desktop/build
-          retention-days: 7
-
-      - name: Upload build artifacts (mac)
-        if: ${{ github.event.inputs.os == 'macos-latest' }}
-        uses: actions/upload-artifact@v4
-        with:
-          name: electron-build-macos-${{ github.event.inputs.mac_arch }}
           path: packages/desktop/build
           retention-days: 7

--- a/.github/workflows/manual-build.yml
+++ b/.github/workflows/manual-build.yml
@@ -40,7 +40,8 @@ jobs:
           github_token: ${{ secrets.github_token }}
           package_root: "packages/desktop"
           release: false  # No release, just build
-          config_overrides: ${{ github.event.inputs.os == 'macos-latest' && format('{"mac": {"target": "default", "arch": "%s"}, "linux": null, "win": null}', github.event.inputs.mac_arch) || '{}' }}
+          Invalid workflow file: .github/workflows/manual-build.yml#L43
+          config_overrides: ${{ github.event.inputs.os == 'macos-latest' && format("{\"mac\": {\"target\": \"default\", \"arch\": \"%s\"}, \"linux\": null, \"win\": null}", github.event.inputs.mac_arch) || '{}' }}
           windows_certs: ${{ secrets.CSC_LINK }}
           windows_certs_password: ${{ secrets.CSC_KEY_PASSWORD }}
 

--- a/.github/workflows/manual-build.yml
+++ b/.github/workflows/manual-build.yml
@@ -4,21 +4,14 @@ on:
   workflow_dispatch:
     inputs:
       os:
-        description: "Operating System / Build target"
+        description: "Operating System"
         required: true
         type: choice
         options:
           - ubuntu-latest
           - windows-latest
           - macos-latest
-      mac_arch:
-        description: "Mac target architecture (only used when os is macos-latest)"
-        required: false
-        default: universal
-        type: choice
-        options:
-          - universal
-          - x64
+          - macos-13
 
 jobs:
   build:
@@ -41,20 +34,13 @@ jobs:
           github_token: ${{ secrets.github_token }}
           package_root: "packages/desktop"
           release: false  # No release, just build
-          # When on macos-latest and mac_arch is set to x64, override the build configuration.
-          config_overrides: ${{ 
-            github.event.inputs.os == 'macos-latest' && github.event.inputs.mac_arch == 'x64'
-              ? '{"mac": {"target": [{"target": "dmg", "arch": ["x64"]}]}}'
-              : '{}' 
-          }}
           windows_certs: ${{ secrets.CSC_LINK }}
           windows_certs_password: ${{ secrets.CSC_KEY_PASSWORD }}
 
       - name: Upload build artifacts
         uses: actions/upload-artifact@v4
         with:
-          name: ${{ github.event.inputs.os == 'macos-latest' && github.event.inputs.mac_arch == 'x64' 
-                    ? 'electron-build-macos-x64' 
-                    : format('electron-build-%s', github.event.inputs.os) }}
-          path: packages/desktop/build
-          retention-days: 7
+          name: electron-build-${{ github.event.inputs.os }}
+          path: |
+            packages/desktop/build
+          retention-days: 7  # Artifacts will be stored for 7 days

--- a/.github/workflows/manual-build.yml
+++ b/.github/workflows/manual-build.yml
@@ -32,22 +32,45 @@ jobs:
         with:
           node-version: 18
 
-      - name: Build/release Electron app
+      - name: Build/release Electron app (non-mac)
+        if: ${{ github.event.inputs.os != 'macos-latest' }}
         uses: AllenCellSoftware/action-electron-builder@fms-file-explorer
         env:
           AMPLITUDE_API_KEY: ${{ secrets.AMPLITUDE_API_KEY }}
         with:
           github_token: ${{ secrets.github_token }}
           package_root: "packages/desktop"
-          release: false  # No release, just build
-          Invalid workflow file: .github/workflows/manual-build.yml#L43
-          config_overrides: ${{ github.event.inputs.os == 'macos-latest' ? format("{\"mac\": {\"target\": \"default\", \"arch\": \"%s\"}, \"linux\": null, \"win\": null}", github.event.inputs.mac_arch) : '{}' }}
+          release: false
           windows_certs: ${{ secrets.CSC_LINK }}
           windows_certs_password: ${{ secrets.CSC_KEY_PASSWORD }}
 
-      - name: Upload build artifacts
+      - name: Build/release Electron app (mac)
+        if: ${{ github.event.inputs.os == 'macos-latest' }}
+        uses: AllenCellSoftware/action-electron-builder@fms-file-explorer
+        env:
+          AMPLITUDE_API_KEY: ${{ secrets.AMPLITUDE_API_KEY }}
+        with:
+          github_token: ${{ secrets.github_token }}
+          package_root: "packages/desktop"
+          release: false
+          # Override configuration to build only mac with the chosen architecture.
+          config_overrides: >
+            {"mac": {"target": "default", "arch": "${{ github.event.inputs.mac_arch }}"}, "linux": null, "win": null}
+          windows_certs: ${{ secrets.CSC_LINK }}
+          windows_certs_password: ${{ secrets.CSC_KEY_PASSWORD }}
+
+      - name: Upload build artifacts (non-mac)
+        if: ${{ github.event.inputs.os != 'macos-latest' }}
         uses: actions/upload-artifact@v4
         with:
-          name: electron-build-${{ github.event.inputs.os }}-${{ github.event.inputs.mac_arch || 'default' }}
+          name: electron-build-${{ github.event.inputs.os }}
+          path: packages/desktop/build
+          retention-days: 7
+
+      - name: Upload build artifacts (mac)
+        if: ${{ github.event.inputs.os == 'macos-latest' }}
+        uses: actions/upload-artifact@v4
+        with:
+          name: electron-build-macos-${{ github.event.inputs.mac_arch }}
           path: packages/desktop/build
           retention-days: 7

--- a/.github/workflows/manual-build.yml
+++ b/.github/workflows/manual-build.yml
@@ -4,7 +4,7 @@ on:
   workflow_dispatch:
     inputs:
       os:
-        description: "Operating System"
+        description: "Operating System / Build target"
         required: true
         type: choice
         options:
@@ -12,13 +12,13 @@ on:
           - windows-latest
           - macos-latest
       mac_arch:
-        description: "Target Mac Architecture (only used for macos-latest)"
+        description: "Mac target architecture (only used when os is macos-latest)"
         required: false
-        default: arm64
+        default: universal
         type: choice
         options:
+          - universal
           - x64
-          - arm64
 
 jobs:
   build:
@@ -33,61 +33,28 @@ jobs:
         with:
           node-version: 18
 
-      # For macOS builds, use one of two steps based on the chosen architecture.
-      - name: Build/release Electron app (mac x64)
-        if: ${{ github.event.inputs.os == 'macos-latest' && github.event.inputs.mac_arch == 'x64' }}
+      - name: Build/release Electron app
         uses: AllenCellSoftware/action-electron-builder@fms-file-explorer
         env:
           AMPLITUDE_API_KEY: ${{ secrets.AMPLITUDE_API_KEY }}
         with:
           github_token: ${{ secrets.github_token }}
           package_root: "packages/desktop"
-          release: false
-          # Override configuration to build only a dmg file for Intel (x64)
-          config_overrides: '{"mac": {"target": [{"target": "dmg", "arch": ["x64"]}]}}'
+          release: false  # No release, just build
+          # When on macos-latest and mac_arch is set to x64, override the build configuration.
+          config_overrides: ${{ 
+            github.event.inputs.os == 'macos-latest' && github.event.inputs.mac_arch == 'x64'
+              ? '{"mac": {"target": [{"target": "dmg", "arch": ["x64"]}]}}'
+              : '{}' 
+          }}
           windows_certs: ${{ secrets.CSC_LINK }}
           windows_certs_password: ${{ secrets.CSC_KEY_PASSWORD }}
 
-      - name: Build/release Electron app (mac arm64)
-        if: ${{ github.event.inputs.os == 'macos-latest' && github.event.inputs.mac_arch == 'arm64' }}
-        uses: AllenCellSoftware/action-electron-builder@fms-file-explorer
-        env:
-          AMPLITUDE_API_KEY: ${{ secrets.AMPLITUDE_API_KEY }}
-        with:
-          github_token: ${{ secrets.github_token }}
-          package_root: "packages/desktop"
-          release: false
-          # Override configuration to build only a dmg file for Apple Silicon (arm64)
-          config_overrides: '{"mac": {"target": [{"target": "dmg", "arch": ["arm64"]}]}}'
-          windows_certs: ${{ secrets.CSC_LINK }}
-          windows_certs_password: ${{ secrets.CSC_KEY_PASSWORD }}
-
-      # For non-macOS builds, use the normal build step.
-      - name: Build/release Electron app (non-mac)
-        if: ${{ github.event.inputs.os != 'macos-latest' }}
-        uses: AllenCellSoftware/action-electron-builder@fms-file-explorer
-        env:
-          AMPLITUDE_API_KEY: ${{ secrets.AMPLITUDE_API_KEY }}
-        with:
-          github_token: ${{ secrets.github_token }}
-          package_root: "packages/desktop"
-          release: false
-          windows_certs: ${{ secrets.CSC_LINK }}
-          windows_certs_password: ${{ secrets.CSC_KEY_PASSWORD }}
-
-      # Upload artifacts appropriately.
-      - name: Upload build artifacts (mac)
-        if: ${{ github.event.inputs.os == 'macos-latest' }}
+      - name: Upload build artifacts
         uses: actions/upload-artifact@v4
         with:
-          name: electron-build-macos-${{ github.event.inputs.mac_arch }}
-          path: packages/desktop/build
-          retention-days: 7
-
-      - name: Upload build artifacts (non-mac)
-        if: ${{ github.event.inputs.os != 'macos-latest' }}
-        uses: actions/upload-artifact@v4
-        with:
-          name: electron-build-${{ github.event.inputs.os }}
+          name: ${{ github.event.inputs.os == 'macos-latest' && github.event.inputs.mac_arch == 'x64' 
+                    ? 'electron-build-macos-x64' 
+                    : format('electron-build-%s', github.event.inputs.os) }}
           path: packages/desktop/build
           retention-days: 7

--- a/.github/workflows/manual-build.yml
+++ b/.github/workflows/manual-build.yml
@@ -11,6 +11,13 @@ on:
           - ubuntu-latest
           - windows-latest
           - macos-latest
+      mac_arch:
+        description: "Target Mac Architecture (only used for macos-latest)"
+        required: false
+        type: choice
+        options:
+          - x64
+          - arm64
 
 jobs:
   build:
@@ -33,13 +40,14 @@ jobs:
           github_token: ${{ secrets.github_token }}
           package_root: "packages/desktop"
           release: false  # No release, just build
+          args: >
+            ${{ github.event.inputs.os == 'macos-latest' && github.event.inputs.mac_arch != '' && ('--mac --' + github.event.inputs.mac_arch) || '' }}
           windows_certs: ${{ secrets.CSC_LINK }}
           windows_certs_password: ${{ secrets.CSC_KEY_PASSWORD }}
 
       - name: Upload build artifacts
         uses: actions/upload-artifact@v4
         with:
-          name: electron-build-${{ github.event.inputs.os }}
-          path: |
-            packages/desktop/build
-          retention-days: 7  # Artifacts will be stored for 7 days
+          name: electron-build-${{ github.event.inputs.os }}-${{ github.event.inputs.mac_arch || 'default' }}
+          path: packages/desktop/build
+          retention-days: 7

--- a/.github/workflows/manual-build.yml
+++ b/.github/workflows/manual-build.yml
@@ -40,8 +40,7 @@ jobs:
           github_token: ${{ secrets.github_token }}
           package_root: "packages/desktop"
           release: false  # No release, just build
-          args: >
-            ${{ github.event.inputs.os == 'macos-latest' && github.event.inputs.mac_arch != '' && format('--mac --{}', github.event.inputs.mac_arch) || '' }}
+          config_overrides: ${{ github.event.inputs.os == 'macos-latest' && format('{"mac": {"target": "default", "arch": "%s"}, "linux": null, "win": null}', github.event.inputs.mac_arch) || '{}' }}
           windows_certs: ${{ secrets.CSC_LINK }}
           windows_certs_password: ${{ secrets.CSC_KEY_PASSWORD }}
 

--- a/.github/workflows/manual-build.yml
+++ b/.github/workflows/manual-build.yml
@@ -41,7 +41,7 @@ jobs:
           package_root: "packages/desktop"
           release: false  # No release, just build
           args: >
-            ${{ github.event.inputs.os == 'macos-latest' && github.event.inputs.mac_arch != '' && ('--mac --' + github.event.inputs.mac_arch) || '' }}
+            ${{ github.event.inputs.os == 'macos-latest' && github.event.inputs.mac_arch != '' && format('--mac --{}', github.event.inputs.mac_arch) || '' }}
           windows_certs: ${{ secrets.CSC_LINK }}
           windows_certs_password: ${{ secrets.CSC_KEY_PASSWORD }}
 

--- a/.github/workflows/manual-build.yml
+++ b/.github/workflows/manual-build.yml
@@ -41,7 +41,7 @@ jobs:
           package_root: "packages/desktop"
           release: false  # No release, just build
           Invalid workflow file: .github/workflows/manual-build.yml#L43
-          config_overrides: ${{ github.event.inputs.os == 'macos-latest' && format("{\"mac\": {\"target\": \"default\", \"arch\": \"%s\"}, \"linux\": null, \"win\": null}", github.event.inputs.mac_arch) || '{}' }}
+          config_overrides: ${{ github.event.inputs.os == 'macos-latest' ? format("{\"mac\": {\"target\": \"default\", \"arch\": \"%s\"}, \"linux\": null, \"win\": null}", github.event.inputs.mac_arch) : '{}' }}
           windows_certs: ${{ secrets.CSC_LINK }}
           windows_certs_password: ${{ secrets.CSC_KEY_PASSWORD }}
 

